test/case/ietf_system/upgrade/test.py

```diff
@@ -53,17 +96,49 @@
             })
 
         with test.step("Wait for upgrade to finish"):
-            for _ in range(600):
+            for left in range(600):
                 oper = target.get_dict("/system-state/software")
                 installer = oper["system-state"]["software"]["installer"]
                 if installer["operation"] == "idle":
                     if "last-error" in installer:
                         print("Install failed:", installer["last-error"])
                         test.fail()
 
-                    test.succeed()
-
+                    break
+                time.sleep(1)
+                if left >= 600:
+                    print("Timeout, last state:", oper)
+                    test.fail()
                 time.sleep(1)
```

Comment:
```suggestion
            else:
                print("Timeout, last state:", oper)
                test.fail()
```
This is a good opportunity to use [for/else](https://book.pythontips.com/en/latest/for_-_else.html), I think :smile: